code/datums/proximity_monitor/fields/heretic_arena.dm

```diff
@@ -0,0 +1,267 @@
+
+/turf/closed/indestructible/heretic_wall/CanAllowThrough(atom/movable/mover, border_dir)
+	if(isliving(mover))
+		var/mob/living/living_mover = mover
+		var/datum/status_effect/arena_tracker/tracker = living_mover.has_status_effect(/datum/status_effect/arena_tracker)
+		if(tracker?.arena_victor)
+			return TRUE
+	return ..()
+
+/turf/closed/indestructible/heretic_wall/Bumped(atom/movable/bumped_atom)
+	. = ..()
+	if(isliving(bumped_atom))
+		var/mob/living/living_mob = bumped_atom
+		var/atom/target = get_edge_target_turf(living_mob, get_dir(src, get_step_away(living_mob, src)))
+		living_mob.throw_at(target, 4, 5)
+		to_chat(living_mob, span_userdanger("The wall repels you with tremendous force!"))
+
+/**
+ * Status applied to every mob in the heretic arena.
+ * Tracks the last person to damage owner.
+ * When owner enters crit, we send a signal to their status so they can leave the arena
+ */
+
+/datum/status_effect/arena_tracker
+	id = "arena_tracker"
+	duration = STATUS_EFFECT_PERMANENT
+	tick_interval = STATUS_EFFECT_NO_TICK
+	status_type = STATUS_EFFECT_UNIQUE
+	alert_type = null
+	/// Tracks the last person who dealt damage to this mob
+	var/mob/last_attacker
+	/// If our mob is free to leave, set to true
+	var/arena_victor = FALSE
+	/// The overlay for our mob, changes color to indicate that they are a victor and are free to leave
+	var/mutable_appearance/crown_overlay
+
+/datum/status_effect/arena_tracker/on_apply()
+	RegisterSignal(owner, SIGNAL_ADDTRAIT(TRAIT_CRITICAL_CONDITION), PROC_REF(on_enter_crit))
+	RegisterSignal(owner, COMSIG_MOB_APPLY_DAMAGE, PROC_REF(damage_taken))
+	RegisterSignal(owner, "COMSIG_OWNER_ENTERED_CRIT", PROC_REF(on_crit_somebody))
+	owner.add_traits(list(TRAIT_ELDRITCH_ARENA_PARTICIPANT, TRAIT_NO_TELEPORT), STATUS_EFFECT_TRAIT)
+	crown_overlay = mutable_appearance('icons/mob/effects/crown.dmi', "arena_fighter", -HALO_LAYER)
+	crown_overlay.pixel_y = 24
+	owner.add_overlay(crown_overlay)
+	return TRUE
+
+/datum/status_effect/arena_tracker/on_remove()
+	UnregisterSignal(owner, list(SIGNAL_ADDTRAIT(TRAIT_CRITICAL_CONDITION), COMSIG_MOB_APPLY_DAMAGE, "COMSIG_OWNER_ENTERED_CRIT"))
+	owner.remove_traits(list(TRAIT_ELDRITCH_ARENA_PARTICIPANT, TRAIT_NO_TELEPORT), STATUS_EFFECT_TRAIT)
+	owner.cut_overlay(crown_overlay)
+	crown_overlay = null
+
+/datum/status_effect/arena_tracker/proc/on_enter_crit(mob/owner)
+	SIGNAL_HANDLER
+	if(!last_attacker) // Safety check in case they somehow enter crit with *nobody* attacking them
+		return
+	SEND_SIGNAL(last_attacker, "COMSIG_OWNER_ENTERED_CRIT")
+
+/datum/status_effect/arena_tracker/proc/damage_taken(
+	datum/source,
+	damage_amount,
+	damagetype,
+	def_zone,
+	blocked,
+	wound_bonus,
+	bare_wound_bonus,
+	sharpness,
+	attack_direction,
+	attacking_item,
+	wound_clothing,
+)
+	SIGNAL_HANDLER
+	if(isnull(attacking_item))
+		stack_trace("proc/damage_taken() was called but without passing attacking_item")
+		return
```

Comment:
This stack trace will happen a lot, an overwhelming majority of damage sources do not pass this. I would just return